### PR TITLE
Tirth patel

### PR DESCRIPTION
**Pull request by Tirth-Patel903**
Hey, Shreya and Shivani I have raised request Now you can review my code.

### CODE REVIEW

File: letter_frequency.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        print("Congratulations You have guesses Correct Word:-",guessed_correctly)
+    else:
+        print("You have used guesses:-", num_of_guesses)
+        print("Game Over Better luck Next Time...")
+    total_game_played=1
+    while True:        
+        ROOT = tk.Tk()
+        ROOT.withdraw()
+    
+        playAgain=messagebox.askquestion("Confirm","Do You Want To Play Again?")
```
Comment (smohan98) [inline letter_frequency.py:33]: MessageBox is a good implementation but from a user's point of view it doesn't work well to see input in one place and output on the terminal

File: letter_frequency.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        print("You have used guesses:-", num_of_guesses)
+        print("Game Over Better luck Next Time...")
+    total_game_played=1
+    while True:        
+        ROOT = tk.Tk()
+        ROOT.withdraw()
+    
+        playAgain=messagebox.askquestion("Confirm","Do You Want To Play Again?")  
+        messagebox_info=["No of Game Played: "+str(total_game_played),"Your Staticals of Game","1st Attempt: "+str(staticals_list[0]),"2nd Attempt: "+str(staticals_list[1]),"3rd Attempt: "+str(staticals_list[2]),"4th Attempt: "+str(staticals_list[3]),"5th Attempt: "+str(staticals_list[4]),"6th Attempt: "+str(staticals_list[5])]
+        staticals_list.pop(0)
```
Comment (smohan98) [inline letter_frequency.py:35]: Rather than using pop 6 times you can just do this : staticals_list[5:]

File: letter_frequency.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            ROOT = tk.Tk()
+            ROOT.withdraw()      
+            messagebox.showinfo("Game Staticals","\n".join(messagebox_info))  
+            break
+        else:
+            print("Goodbye")
+
+def word_fetch():
+    try:
+        a = open("demo.txt", "w")
```
Comment (smohan98) [inline letter_frequency.py:57]: For code to be readable for other developers use better naming convention

File: letter_frequency.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        print("File not accessible")
+    a.close()
+    for key,value in freq.items():
+        oitem=[]
+        ritem=value
+        for item in ritem:
+            oitem.append(item/count)
+        rank[key]=oitem
+    sorted_list = sorted(freq.items(), key=lambda x:x[1])
+    print(sorted_list)
```
Comment (smohan98) [inline letter_frequency.py:103]: Unnecessary print statements should not be a part of this PR

File: letter_frequency.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            writer.writerow([key, value])
+    listOfTuples(freq)
+
+#function for list of dictionary into tuple of dictionary
+def listOfTuples(freq):
+    tupleDict={}
+    for key, value in freq.items():
+        tuple(value)
+        tupleDict[key]=tuple(value)
+    print(tupleDict)
```
Comment (smohan98) [inline letter_frequency.py:123]: Unnecessary print statements should not be a part of this PR

File: letter_frequency.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        else:
+            clue += "_" #_ represents False Letter 
+        position += 1
+    print(clue)
+    staticals_list.append(distribution)
+    print(staticals_list)
+    logger.info('Your clue:-'+ clue)
+    logger.info('Answer:-'+ theAnswer)
+    logger.info('Your Guess:-'+ theGuess)
+    logger.info('Your guessing percentage:-'+str(distribution))
```
Comment (smohan98) [inline letter_frequency.py:145]: Logs are a developer facing entity and should be written in a correct format
 ```
User Input :: INFO  :: info message
 User Input :: WARNING  :: warn message
 User Input :: ERROR    :: error message
 User Input :: CRITICAL :: critical message
```

File: test_wordle.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    def test_process(self):
+        guess="earth"
+        answer="world"
+        self.assertEqual(WordleGame.processGuess(self,guess,answer),False,"test_process Test failed")
+
+    #testing function for same guess
+    def test_process_same_guess(self):
+        guess="earth"
+        answer="earth"
+        self.assertEqual(WordleGame.processGuess(self,guess,answer),True,"test_process Test failed")
```
Comment (smohan98) [inline test_wordle.py:16]:
```suggestion
        self.assertEqual(WordleGame.processGuess(self,guess,answer),True,"test_process_same_guess Test failed")

```

File: wordle.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            print("You have used guesses:-", num_of_guesses)
+            print("Game Over Better luck Next Time...")
+        total_game_played=1
+        #logic for stataticals
+        while True:        
+            ROOT = tk.Tk()
+            ROOT.withdraw()
+            playAgain=messagebox.askquestion("Confirm","Do You Want To Play Again?")  
+            messagebox_info=["No of Game Played: "+str(total_game_played),"Your Staticals of Game","1st Attempt: "+str(staticals_list[0]),"2nd Attempt: "+str(staticals_list[1]),"3rd Attempt: "+str(staticals_list[2]),"4th Attempt: "+str(staticals_list[3]),"5th Attempt: "+str(staticals_list[4]),"6th Attempt: "+str(staticals_list[5])]
+            staticals_list.pop(0)
```
Comment (smohan98) [inline wordle.py:36]: Rather than using pop 6 times you can just do this : staticals_list[5:]

File: wordle.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            logger.info(messagebox_info)
+            if playAgain == 'yes':
```
Comment (smohan98) [inline wordle.py:43]: Boolean Flags supersed any other datatype flags, its good practice to use Booleans

File: wordle.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+            print("File not accessible")
+        a.close()
+        f.close()
+
+
+    def processGuess(self,theAnswer, theGuess):
+        position=0
+        clue=""
+        hint=""
+        pos=""
```
Comment (smohan98) [inline wordle.py:73]:
```suggestion
        clue,hint,pos="","",""

```

File: wordle.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        distribution=0
+        for letter in theGuess:
+            if letter == theAnswer[position]:
+                hint += letter + ''
+                pos=str(position) +','
```
Comment (smohan98) [inline wordle.py:78]:
```suggestion
                pos=f"{position},"

```

File: wordle.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+            array_hintt.append(i)
+        linked_list = collections.deque()
+        line_count=0
+        if(hint!=""):
+            print(len(hint))
+            wordList=open("demo.txt", "r+")
+            lines = wordList.read().splitlines()
+            for line in lines:
+                for i in array_pos:
+                    if (line[0] == theAnswer[0]) and (line[4] == theAnswer[4]):
```
Comment (smohan98) [inline wordle.py:119]:
```suggestion
                    if line[0] == theAnswer[0] and line[4] == theAnswer[4]:

```

File: wordle.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+                        linked_list.append(line)
+                    elif(hint in line):
+                        linked_list.append(line)
+                        
+            print(linked_list)
```
Comment (smohan98) [inline wordle.py:124]: Unnecessary print statements should not be a part of this PR

File: wordle.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+        logger.info('Your Guess:-'+ theGuess)
+        logger.info('Your guessing percentage:-'+str(distribution))
+        return clue == "GGGGG" #True if correct,False otherwie
+
+#verifies the word is valid or not which is guessed by user
+    def verifyGuess(self,theGuess):
+        if(len(theGuess) > 5 ):
+            return theGuess[:5].strip()
+        else:
+            return theGuess
```
Comment (smohan98) [inline wordle.py:102]:
```suggestion
        return theGuess

```

File: letter_frequency.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+                            citem[0]+=1
+                        elif j==1:
+                            citem[1]+=1
+                        elif j==2:
+                            citem[2]+=1
+                        elif j==3:
+                            citem[3]+=1
+                        elif j==4:
+                            citem[4]+=1
+                        freq[item]=citem
```
Comment (smohan98) [inline letter_frequency.py:90]: So many if else is a bruteforce solution and can easily be avoided by a dictionary implementation

File: letter_frequency.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+    wordList=open("demo.txt", "r+")
+    words = wordList.read().splitlines()
+    answer=random.choice(words)
+    removeWord(answer)
+    num_of_guesses=0
+    guesses_correctly = False
+    print("G=Letter is True, Y=True Letter In wrong Spot and _= Wrong Letter")
+    while num_of_guesses < 6 and not guesses_correctly:
+        #get input from user
+        ROOT = tk.Tk()
```
Comment (smohan98) [inline letter_frequency.py:180]: Can be replaced by from tkinter import TK

File: letter_frequency.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+    logger.info('Answer:-'+ theAnswer)
+    logger.info('Your Guess:-'+ theGuess)
+    logger.info('Your guessing percentage:-'+str(distribution))
+    return clue == "GGGGG" #True if correct,False otherwie
+
+def verifyGuess(theGuess):
+    if(len(theGuess) > 5 ):
+        return theGuess[:5].strip()
+    else:
+        return theGuess
```
Comment (smohan98) [inline letter_frequency.py:152]:
```suggestion
    return theGuess
```

**Review by smohan98 — COMMENTED** [review]
Following are some comments and suggestions I have about this PR:

- Most importantly the code is not at all segregated in files, 1 file code is a very unacceptable way to write code by industry standards
- No function has a docstring
- Code is not readable for a new developer to understand easily
- File names in python don't follow camel casing, they should follow '_' separated names
- Too many unnecessary print statements in the code
- Message box implementation not a good idea in this case as user will be forced to have two windows open( i.e. the terminal for result and input box for input)
- Logging needs to be improved to understand errors and pinpoint problems if occurred
- Variable naming should not be single letters but rather should be a self descriptive
- Boolean flags should be used for conditional loops and not 'yes' and 'no'
- Majority functions lack try catch loops to catch errors if any
- Test file should have better messages for test failures
- Some of the code follows Python2 standards and some follows Python3
- Too many Bruteforce solutions are implemented

Pros:
- Although the use-case doesn't suit here, input box is an impressive implementation
- Comments are present before some logics
-  Functions are present for even small small cases to be reused

File: wordle.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+        try:
+            a = open("demo.txt", "w")
+            f = open("words.txt", "r")
+            for x in f:
+                if(len(x)==6):
+                    a.write(x)
+        except IOError:
+            print("File not accessible")
+        a.close()
+        f.close()
```
Comment (smaurya30) [inline wordle.py:66]: close() function can be added in finally blocks in case an exception occurs

File: letter_frequency.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+#Let us Create an object 
+logger=logging.getLogger() 
+
+#Now we are going to Set the threshold of logger to DEBUG 
+logger.setLevel(logging.DEBUG)
+todayDate=datetime.today().strftime("%m/%d/%Y, %H:%M:%S")
+logger.info('Your Gameplay for '+todayDate)
+
+staticals_list=[]
+def main():
```
Comment (smaurya30) [inline letter_frequency.py:20]: Functions need to be added to a class for OOP requirement

File: test_wordle.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
+import unittest
+from wordle import WordleGame
+
+class Test_Wordle(unittest.TestCase):
+
+    # function which check word is matched or not
+    def test_process(self):
```
Comment (smaurya30) [inline test_wordle.py:7]: Good job on the test cases, most of them are covered

File: letter_frequency.py (hunk 1 above)
Comment (smaurya30) [reply letter_frequency.py:33]: Agree with @smohan98

File: wordle.py [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+            ROOT = tk.Tk()
+            ROOT.withdraw()
+            playAgain=messagebox.askquestion("Confirm","Do You Want To Play Again?")  
+            messagebox_info=["No of Game Played: "+str(total_game_played),"Your Staticals of Game","1st Attempt: "+str(staticals_list[0]),"2nd Attempt: "+str(staticals_list[1]),"3rd Attempt: "+str(staticals_list[2]),"4th Attempt: "+str(staticals_list[3]),"5th Attempt: "+str(staticals_list[4]),"6th Attempt: "+str(staticals_list[5])]
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
+            staticals_list.pop(0)
```
Comment (smaurya30) [inline wordle.py:41]: Can use clear() instead of pop() 6 times for emptying the list

File: wordle.py (hunk 9 above)
Comment (smaurya30) [reply wordle.py:43]: Agree with @smohan98

File: wordle.py [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+            position += 1
+        self.find_similar_words(hint,clue,pos,theAnswer)
+        staticals_list.append(distribution)
+        print(staticals_list)
+        print(clue)
+        logger.info('Your clue:-'+ clue)
+        logger.info('Answer:-'+ theAnswer)
+        logger.info('Your Guess:-'+ theGuess)
+        logger.info('Your guessing percentage:-'+str(distribution))
+        return clue == "GGGGG" #True if correct,False otherwie
```
Comment (smaurya30) [inline wordle.py:95]: Missing Exception Handling

File: wordle.py [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+#verifies the word is valid or not which is guessed by user
+    def verifyGuess(self,theGuess):
+        if(len(theGuess) > 5 ):
+            return theGuess[:5].strip()
+        else:
+            return theGuess
+
+#module for finding words which are similars to guessed word
+    def find_similar_words(self,hint,clue,pos,theAnswer):
+        array_hintt=[]
```
Comment (smaurya30) [inline wordle.py:106]: Missing Exception Handling

**Review by smaurya30 — CHANGES_REQUESTED** [review]

File: letter_frequency.py (hunk 1 above)
Comment (Tirth-Patel903) [reply letter_frequency.py:33]: Yes, but I want GUI best so this is implemented by me. Thanks for suggestions.

File: letter_frequency.py (hunk 2 above)
Comment (Tirth-Patel903) [reply letter_frequency.py:35]: I done this change you can review it, Thanks.

File: letter_frequency.py (hunk 3 above)
Comment (Tirth-Patel903) [reply letter_frequency.py:57]: changed this you can check

File: letter_frequency.py (hunk 4 above)
Comment (Tirth-Patel903) [reply letter_frequency.py]: removed this print statement you can review.

File: wordle.py (hunk 8 above)
Comment (Tirth-Patel903) [reply wordle.py]: This changed to suggested you can check.

File: wordle.py (hunk 9 above)
Comment (Tirth-Patel903) [reply wordle.py:38]: This is inbuild functionality for boolean there is another function in tkinter but I used this . Thanks but can't change.

File: wordle.py (hunk 10 above)
Comment (Tirth-Patel903) [reply wordle.py]: changed suggested code thanks.

File: wordle.py (hunk 11 above)
Comment (Tirth-Patel903) [reply wordle.py:71]: changed suggested code, thanks.

File: wordle.py [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+        array_hint=hint.split()
+        array_pos=pos.split(",")
+        for i in hint:
+            array_hintt.append(i)
+        linked_list = collections.deque()
+        line_count=0
+        if(hint!=""):
+            print(len(hint))
```
[the remark this answers is not in the document]
Comment (Tirth-Patel903) [reply wordle.py]: removed this thanks

**Pull request merged**